README.md
go-rsql
=======

# overview
RSQL is a query language for parametrized filtering of entries in APIs. 
It is based on FIQL (Feed Item Query Language) – an URI-friendly syntax for expressing filters across the entries in an Atom Feed.
FIQL is great for use in URI; there are no unsafe characters, so URL encoding is not required.
On the other side, FIQL’s syntax is not very intuitive and URL encoding isn’t always that big deal,
so RSQL also provides a friendlier syntax for logical operators and some comparison operators.

This is a small RSQL helper library, written in golang.
It can be used to parse a RSQL string and turn it into a database query string.

Currently, only mongodb is supported out of the box (however it is very easy to extend the parser if needed).

# basic usage
```go
package main

import (

"github.com/rbicker/go-rsql"
"log"
)

func main(){
	parser, err := rsql.NewParser(rsql.Mongo())
	if err != nil {
		log.Fatalf("error while creating parser: %s", err)
	}
	s := `status=="A",qty=lt=30`
	res, err := parser.Process(s)
	if err != nil {
		log.Fatalf("error while parsing: %s", err)
	}
	log.Println(res)
	// { "$or": [ { "status": "A" }, { "qty": { "$lt": 30 } } ] }
}
```


# supported operators
The library supports the following basic operators by default:

| Basic Operator | Description         |
|----------------|---------------------|
| ==             | Equal To            |
| !=             | Not Equal To        |
| =gt=           | Greater Than        |
| =ge=           | Greater Or Equal To |
| =lt=           | Less Than           |
| =le=           | Less Or Equal To    |
| =in=           | In                  |
| =out=          | Not in              |

The following table lists two joining operators:

| Composite Operator | Description         |
|--------------------|---------------------|
| ;                  | Logical AND         |
| ,                  | Logical OR          |


# advanced usage 

## custom operators
The library makes it easy to define custom operators:
```go
package main

import (

"fmt"
"github.com/rbicker/go-rsql"
"log"
)

func main(){
    // create custom operators for "exists"- and "all"-operations
    customOperators := []rsql.Operator{
        {
            Operator:       "=ex=",
            Formatter: func (key, value string) string {
                return fmt.Sprintf(`{ "%s": { "$exists": %s } }`, key, value)
            },
        },
        {
            Operator:       "=all=",
            Formatter: func(key, value string) string {
                return fmt.Sprintf(`{ "%s": { "$all": [ %s ] } }`, key, value[1:len(value)-1])
            },
        },
    }
    // create parser with default mongo operators
    // plus the two custom operators
    var opts []func(*rsql.Parser) error
    opts = append(opts, rsql.Mongo())
    opts = append(opts, rsql.WithOperators(customOperators...))
	parser, err := rsql.NewParser(opts...)
	if err != nil {
		log.Fatalf("error while creating parser: %s", err)
	}
    // parse string with some default operators
    res, err := parser.Process(`(a==1;b==2),c=gt=5`)
	if err != nil {
		log.Fatalf("error while parsing: %s", err)
	}
	log.Println(res)
	// { "$or": [ { "$and": [ { "a": 1 }, { "b": 2 } ] }, { "c": { "$gt": 5 } } ] }
    
    // use custom operator =ex=
	res, err = parser.Process(`a=ex=true`)
	if err != nil {
		log.Fatalf("error while parsing: %s", err)
	}
	log.Println(res)
	// { "a": { "$exists": true } }
    
    // use custom list operator =all=
	res, err = parser.Process(`tags=all=('waterproof','rechargeable')`)
	if err != nil {
		log.Fatalf("error while parsing: %s", err)
	}
	log.Println(res)
	// { "tags": { "$all": [ 'waterproof','rechargeable' ] } }
}
```

## transform keys
If your database key naming scheme is different from the one used in your rsql statements, you can add functions to transform your keys.

```go
package main

import (
	"github.com/rbicker/go-rsql"
	"log"
	"strings"
)

func main() {
	transformer := func(s string) string {
		return strings.ToUpper(s)
	}
	parser, err := rsql.NewParser(rsql.Mongo(), rsql.WithKeyTransformers(transformer))
	if err != nil {
		log.Fatalf("error while creating parser: %s", err)
	}
	s := `status=="a",qty=lt=30`
	res, err := parser.Process(s)
	if err != nil {
		log.Fatalf("error while parsing: %s", err)
	}
	log.Println(res)
	// { "$or": [ { "STATUS": "a" }, { "QTY": { "$lt": 30 } } ] }
}
```

## define allowed or forbidden keys
```go
package main

import (
	"github.com/rbicker/go-rsql"
	"log"
)

func main() {
	parser, err := rsql.NewParser(rsql.Mongo())
	if err != nil {
		log.Fatalf("error while creating parser: %s", err)
	}
	s := `status=="a",qty=lt=30`
	_, err = parser.Process(s, rsql.SetAllowedKeys([]string{"status, qty"}))
	// -> ok
	_, err = parser.Process(s, rsql.SetAllowedKeys([]string{"status"}))
	// -> error
	_, err = parser.Process(s, rsql.SetForbiddenKeys([]string{"status"}))
	// -> error
	_, err = parser.Process(s, rsql.SetAllowedKeys([]string{"age"}))
	// -> ok
}
```

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.4.0] - 2021-08-01
### Changed
* the equal query operator for mongodb is formatted as { "key": value } instead of { "key": { "$eq": value } } to provide more flexibility 

## [0.3.0] - 2021-02-17
### Added
* added process options to define to allow or disallow specific keys.
* added key transformers.

## [0.2.0] - 2020-05-29
### Changed
* mongodb related logical and query operators are now provided as "formatter-functions".
This makes the parser much more flexible, however the `Mongo()`-function needs to be passed to `NewParser()` from now on. 

## [0.1.0] - 2020-05-28
### Changed
* custom operators can now define MongoFormatter as a function instead of a MongoOperator string to increase the flexibility
### Fixed
* allow nested parentheses, so it is possible to define filters like: userIds=in=(ObjectId("xxx"),ObjectId("yyy"))

## [0.0.2] - 2020-05-23
### Added
* added the ability to handle empty rsql strings

## [0.0.1] - 2020-05-16
### Added
* parser to parse rsql string to mongodb json filter string 
* functionality to add custom operators

rsql_test.go
package rsql

import (
	"fmt"
	"reflect"
	"strings"
	"testing"
)

func Test_encodeSpecial(t *testing.T) {
	tests := []struct {
		name string
		s    string
		want string
	}{
		{
			name: "encode special strings",
			s:    `(x==1;y=='\(hello\, how are you?\;thanks\)')`,
			want: "(x==1;y=='%5C%28hello%5C%2C how are you?%5C%3Bthanks%5C%29')",
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := encodeSpecial(tt.s); got != tt.want {
				t.Errorf("encodeSpecial() = %v, want %v", got, tt.want)
			}
		})
	}
}

func Test_decodeSpecial(t *testing.T) {
	tests := []struct {
		name string
		s    string
		want string
	}{
		{
			name: "decode special strings",
			s:    "(x==1;y=='%5C%28hello%5C%2C how are you?%5C%3Bthanks%5C%29')",
			want: `(x==1;y=='\(hello\, how are you?\;thanks\)')`,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := decodeSpecial(tt.s); got != tt.want {
				t.Errorf("decodeSpecial() = %v, want %v", got, tt.want)
			}
		})
	}
}

func Test_findOuterParentheses(t *testing.T) {
	tests := []struct {
		name    string
		s       string
		n       int
		want    [][]int
		wantErr bool
	}{
		{
			name:    "parentheses not matching",
			s:       "(x)(a(b)",
			wantErr: true,
		},
		{
			name: "one",
			s:    "(x==1)",
			want: [][]int{
				{
					0,
					5,
				},
			},
		},
		{
			name: "two",
			s:    "(x==1),(y==1)",
			want: [][]int{
				{
					0,
					5,
				},
				{
					7,
					12,
				},
			},
		},
		{
			name: "object id",
			s:    `(a==1),(_id=in=(ObjectId("xxx"),ObjectId("yyy")))`,
			want: [][]int{
				{
					0,
					5,
				},
				{
					7,
					48,
				},
			},
		},
		{
			name: "two, but only look for one",
			s:    "(x==1),(y==1)",
			n:    1,
			want: [][]int{
				{
					0,
					5,
				},
			},
		},
		{
			name: "containing list",
			s:    "(y==2),x=in=(1,2,3)",
			want: [][]int{
				{
					0,
					5,
				},
			},
		},
		{
			name: "containing nested list",
			s:    "(y==2),x=in=(1,(2),3)",
			want: [][]int{
				{
					0,
					5,
				},
			},
		},
		{
			name: "at beginning",
			s:    "(b==1,c==1);a==1",
			want: [][]int{
				{
					0,
					10,
				},
			},
		},
		{
			name: "at end",
			s:    "a==1;(b==1,c==1)",
			want: [][]int{
				{
					5,
					15,
				},
			},
		},
		{
			name: "nested",
			s:    "((a==1,a==2),(b==1,b==2))",
			want: [][]int{
				{
					0,
					24,
				},
			},
		},
		{
			name: "nested part 2",
			s:    "(a==1,a==2),(b==1,b==2)",
			want: [][]int{
				{
					0,
					10,
				},
				{
					12,
					22,
				},
			},
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := findOuterParentheses(tt.s, tt.n)
			if (err != nil) != tt.wantErr {
				t.Errorf("findOuterParentheses() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if !reflect.DeepEqual(got, tt.want) {
				t.Errorf("findOuterParentheses() got = %v, want %v", got, tt.want)
			}
		})
	}
}

func TestParser_ProcessMongo(t *testing.T) {
	tests := []struct {
		name            string
		s               string
		options         []func(*ProcessOptions) error
		customOperators []Operator
		keyTransformers []func(s string) string
		want            string
		wantErr         bool
	}{
		{
			name: "empty",
			s:    "",
			want: "{ }",
		},
		{
			name: "==",
			s:    "a==1",
			want: `{ "a": 1 }`,
		},
		{
			name: "!=",
			s:    "a!=1",
			want: `{ "a": { "$ne": 1 } }`,
		},
		{
			name: "=gt=",
			s:    "a=gt=1",
			want: `{ "a": { "$gt": 1 } }`,
		},
		{
			name: "=ge=",
			s:    "a=ge=1",
			want: `{ "a": { "$gte": 1 } }`,
		},
		{
			name: "=lt=",
			s:    "a=lt=1",
			want: `{ "a": { "$lt": 1 } }`,
		},
		{
			name: "=le=",
			s:    "a=le=1",
			want: `{ "a": { "$lte": 1 } }`,
		},
		{
			name: "=in=",
			s:    "a=in=(1,2,3)",
			want: `{ "a": { "$in": 1,2,3 } }`,
		},
		{
			name: "=out=",
			s:    "a=out=(1,2,3)",
			want: `{ "a": { "$nin": 1,2,3 } }`,
		},
		{
			name: "(a==1)",
			s:    "(a==1)",
			want: `{ "a": 1 }`,
		},
		{
			name: "a==1;b==2",
			s:    "a==1;b==2",
			want: `{ "$and": [ { "a": 1 }, { "b": 2 } ] }`,
		},
		{
			name: "a==1,b==2",
			s:    "a==1,b==2",
			want: `{ "$or": [ { "a": 1 }, { "b": 2 } ] }`,
		},
		{
			name: "a==1;b==2,c==1",
			s:    "a==1;b==2,c==1",
			want: `{ "$or": [ { "$and": [ { "a": 1 }, { "b": 2 } ] }, { "c": 1 } ] }`,
		},
		{
			name: "(a==1;b==2),c=gt=5",
			s:    "(a==1;b==2),c=gt=5",
			want: `{ "$or": [ { "$and": [ { "a": 1 }, { "b": 2 } ] }, { "c": { "$gt": 5 } } ] }`,
		},
		{
			name: "c==1,(a==1;b==2)",
			s:    "c==1,(a==1;b==2)",
			want: `{ "$or": [ { "c": 1 }, { "$and": [ { "a": 1 }, { "b": 2 } ] } ] }`,
		},
		{
			name: "a==1;(b==1,c==2)",
			s:    "a==1;(b==1,c==2)",
			want: `{ "$and": [ { "a": 1 }, { "$or": [ { "b": 1 }, { "c": 2 } ] } ] }`,
		},
		{
			name: "(a==1,b==1);(c==1,d==2)",
			s:    "(a==1,b==1);(c==1,d==2)",
			want: `{ "$and": [ { "$or": [ { "a": 1 }, { "b": 1 } ] }, { "$or": [ { "c": 1 }, { "d": 2 } ] } ] }`,
		},
		{
			name: "custom operator: =ex=",
			s:    "a=ex=true",
			customOperators: []Operator{
				{
					Operator: "=ex=",
					Formatter: func(key, value string) string {
						return fmt.Sprintf(`{ "%s": { "$exists": %s } }`, key, value)
					},
				},
			},
			want: `{ "a": { "$exists": true } }`,
		},
		{
			name: "custom list operator: =all=",
			s:    "tags=all=('waterproof','rechargeable')",
			customOperators: []Operator{
				{
					Operator: "=all=",
					Formatter: func(key, value string) string {

						return fmt.Sprintf(`{ "%s": { "$all": [ %s ] } }`, key, value[1:len(value)-1])
					},
				},
			},
			want: `{ "tags": { "$all": [ 'waterproof','rechargeable' ] } }`,
		},
		{
			name:    "all keys allowed",
			s:       "a==1",
			options: []func(*ProcessOptions) error{},
			wantErr: false,
			want:    `{ "a": 1 }`,
		},
		{
			name: "key allowed",
			s:    "a==1",
			options: []func(*ProcessOptions) error{
				SetAllowedKeys([]string{"a"}),
			},
			wantErr: false,
			want:    `{ "a": 1 }`,
		},
		{
			name: "key not allowed",
			s:    "a==1",
			options: []func(*ProcessOptions) error{
				SetAllowedKeys([]string{"b"}),
			},
			wantErr: true,
			want:    "",
		},
		{
			name: "key forbidden",
			s:    "a==1",
			options: []func(*ProcessOptions) error{
				SetForbiddenKeys([]string{"a"}),
			},
			wantErr: true,
			want:    "",
		},
		{
			name: "key not forbidden",
			s:    "a==1",
			options: []func(*ProcessOptions) error{
				SetForbiddenKeys([]string{"b"}),
			},
			wantErr: false,
			want:    `{ "a": 1 }`,
		},
		{
			name: "uppercase key transformer",
			s:    "a==1",
			keyTransformers: []func(s string) string{
				func(s string) string {
					return strings.ToUpper(s)
				},
			},
			wantErr: false,
			want:    `{ "A": 1 }`,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			var opts []func(*Parser) error
			opts = append(opts, Mongo())
			opts = append(opts, WithOperators(tt.customOperators...))
			opts = append(opts, WithKeyTransformers(tt.keyTransformers...))
			parser, err := NewParser(opts...)
			if err != nil {
				t.Fatalf("error while creating parser: %s", err)
			}
			got, err := parser.Process(tt.s, tt.options...)
			if (err != nil) != tt.wantErr {
				t.Errorf("Process() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if got != tt.want {
				t.Errorf("Process() got = %v, want %v", got, tt.want)
			}
		})
	}
}

func Test_findParts(t *testing.T) {
	tests := []struct {
		name       string
		s          string
		n          int
		separators []string
		want       [][]int
		wantErr    bool
	}{
		{
			name: "empty",
			s:    "",
			n:    -1,
		},
		{
			name:    "no separators",
			s:       "(a==1),(b==1)",
			n:       -1,
			wantErr: true,
		},
		{
			name:       "start with separators",
			s:          ",(a==1),(b==1)",
			n:          -1,
			separators: []string{","},
			wantErr:    true,
		},
		{
			name:       "end with separators",
			s:          "(a==1),(b==1),",
			n:          -1,
			separators: []string{","},
			wantErr:    true,
		},
		{
			name:       "parentheses mismatch",
			s:          "(a==1)),(b==1)",
			n:          -1,
			separators: []string{","},
			wantErr:    true,
		},
		{
			name:       "parentheses mismatch in operation",
			s:          "a=in=(1),2,3)",
			n:          -1,
			separators: []string{","},
			wantErr:    true,
		},
		{
			name:       "nested",
			s:          "((a==1),(b==1)),(c==1)",
			n:          -1,
			separators: []string{","},
			want: [][]int{
				{
					0,
					15,
				},
				{
					16,
					22,
				},
			},
		},
		{
			name:       "return one",
			s:          "(a==1),(b==1)",
			n:          1,
			separators: []string{","},
			want: [][]int{
				{
					0,
					6,
				},
			},
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := findParts(tt.s, tt.n, tt.separators...)
			if (err != nil) != tt.wantErr {
				t.Errorf("findParts = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if !reflect.DeepEqual(got, tt.want) {
				t.Errorf("findParts() got = %v, want %v", got, tt.want)
			}
		})
	}
}

func Test_findORs(t *testing.T) {
	tests := []struct {
		name    string
		s       string
		n       int
		want    [][]int
		wantErr bool
	}{
		{
			name: "simple or",
			s:    "(a==1),(b==1)",
			n:    -1,
			want: [][]int{
				{
					0,
					6,
				},
				{
					7,
					13,
				},
			},
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := findORs(tt.s, tt.n)
			if (err != nil) != tt.wantErr {
				t.Errorf("findORs() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if !reflect.DeepEqual(got, tt.want) {
				t.Errorf("findORs() got = %v, want %v", got, tt.want)
			}
		})
	}
}

func Test_findANDs(t *testing.T) {
	tests := []struct {
		name    string
		s       string
		n       int
		want    [][]int
		wantErr bool
	}{
		{
			name: "simple and",
			s:    "(a==1);(b==1)",
			n:    -1,
			want: [][]int{
				{
					0,
					6,
				},
				{
					7,
					13,
				},
			},
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := findANDs(tt.s, tt.n)
			if (err != nil) != tt.wantErr {
				t.Errorf("findANDs() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if !reflect.DeepEqual(got, tt.want) {
				t.Errorf("findANDs() got = %v, want %v", got, tt.want)
			}
		})
	}
}


rsql.go
package rsql

import (
	"fmt"
	"regexp"
	"strings"
)

// specialEncode is the map for encoding
// a list of special characters that could
// mess up the parser.
var specialEncode = map[string]string{
	`\(`: "%5C%28",
	`\)`: "%5C%29",
	`\,`: "%5C%2C",
	`\;`: "%5C%3B",
	`\=`: "%5C%3D",
}

// regex to match Operator within operation
var reOperator = regexp.MustCompile(`([!=])[^=()]*=`)

// Operator represents a query Operator.
// It defines the Operator itself, the mongodb representation
// of the Operator and if it is a list Operator or not.
// Operators must match regex reOperator: `(!|=)[^=()]*=`
type Operator struct {
	Operator  string
	Formatter func(key, value string) string
}

// Parser represents a RSQL parser.
type Parser struct {
	operators       []Operator
	andFormatter    func(ss []string) string
	orFormatter     func(ss []string) string
	keyTransformers []func(s string) string
}

// NewParser returns a new rsql server.
func NewParser(options ...func(*Parser) error) (*Parser, error) {
	// create parser
	var parser = Parser{}
	// run functional options
	for _, op := range options {
		err := op(&parser)
		if err != nil {
			return nil, fmt.Errorf("setting option failed: %w", err)
		}
	}
	if parser.andFormatter == nil {
		return nil, fmt.Errorf("AND-formatter is not defined")
	}
	if parser.orFormatter == nil {
		return nil, fmt.Errorf("OR-formatter is not defined")
	}
	return &parser, nil
}

// Mongo adds the default mongo operators to the parser
func Mongo() func(parser *Parser) error {
	return func(parser *Parser) error {
		// operators
		var operators = []Operator{
			{
				"==",
				func(key, value string) string {
					return fmt.Sprintf(`{ "%s": %s }`, key, value)
				},
			},
			{
				"!=",
				func(key, value string) string {
					return fmt.Sprintf(`{ "%s": { "$ne": %s } }`, key, value)
				},
			},
			{
				"=gt=",
				func(key, value string) string {
					return fmt.Sprintf(`{ "%s": { "$gt": %s } }`, key, value)
				},
			},
			{
				"=ge=",
				func(key, value string) string {
					return fmt.Sprintf(`{ "%s": { "$gte": %s } }`, key, value)
				},
			},
			{
				"=lt=",
				func(key, value string) string {
					return fmt.Sprintf(`{ "%s": { "$lt": %s } }`, key, value)
				},
			},
			{
				"=le=",
				func(key, value string) string {
					return fmt.Sprintf(`{ "%s": { "$lte": %s } }`, key, value)
				},
			},
			{
				"=in=",
				func(key, value string) string {
					// remove parentheses
					value = value[1 : len(value)-1]
					return fmt.Sprintf(`{ "%s": { "$in": %s } }`, key, value)
				},
			},
			{
				"=out=",
				func(key, value string) string {
					// remove parentheses
					value = value[1 : len(value)-1]
					return fmt.Sprintf(`{ "%s": { "$nin": %s } }`, key, value)
				},
			},
		}
		parser.operators = append(parser.operators, operators...)
		// AND formatter
		parser.andFormatter = func(ss []string) string {
			if len(ss) > 1 {
				return fmt.Sprintf(`{ "$and": [ %s ] }`, strings.Join(ss, ", "))
			}
			if len(ss) == 0 {
				return ""
			}
			return ss[0]
		}
		// OR formatter
		parser.orFormatter = func(ss []string) string {
			if len(ss) > 1 {
				return fmt.Sprintf(`{ "$or": [ %s ] }`, strings.Join(ss, ", "))
			}
			if len(ss) == 0 {
				return "{ }"
			}
			return ss[0]
		}
		return nil
	}
}

// WithOperator adds custom operators to the parser
func WithOperators(operators ...Operator) func(parser *Parser) error {
	return func(parser *Parser) error {
		for _, o := range operators {
			if !reOperator.MatchString(o.Operator) {
				return fmt.Errorf("invalid Operator '%s' as it does not match regex `(!|=)[^=]*=`", o.Operator)
			}
		}
		parser.operators = append(parser.operators, operators...)
		return nil
	}
}

// WithKeyTransformers adds functions to alter key names in any way.
func WithKeyTransformers(transformers ...func(string) string) func(parser *Parser) error {
	return func(parser *Parser) error {
		parser.keyTransformers = append(parser.keyTransformers, transformers...)
		return nil
	}
}

// ProcessOptions contains options for the parser's Process function.
type ProcessOptions struct {
	allowedKeys   []string
	forbiddenKeys []string
}

// SetAllowedKeys set's the keys which can be used for querying.
func SetAllowedKeys(keys []string) func(opts *ProcessOptions) error {
	return func(opts *ProcessOptions) error {
		opts.allowedKeys = keys
		return nil
	}
}

// SetForbiddenKeys set's the keys which must not be used for querying.
func SetForbiddenKeys(keys []string) func(opts *ProcessOptions) error {
	return func(opts *ProcessOptions) error {
		opts.forbiddenKeys = keys
		return nil
	}
}

// containsString checks if a given slice of strings contains a given string.
func containsString(ss []string, s string) bool {
	for _, x := range ss {
		if x == s {
			return true
		}
	}
	return false
}

// Process takes the given string and processes it using parser's operators.
func (parser *Parser) Process(s string, options ...func(*ProcessOptions) error) (string, error) {
	// set process options
	opts := ProcessOptions{}
	for _, op := range options {
		err := op(&opts)
		if err != nil {
			return "", fmt.Errorf("setting process option failed: %w", err)
		}
	}
	// regex to match identifier within operation, before the equal or expression mark
	var reKey = regexp.MustCompile(`^[^=!]+`)
	// regex to match value within the operation, after the equal sign
	var reValue = regexp.MustCompile(`[^=]+$`)
	// get ORs
	locations, err := findORs(s, -1)
	if err != nil {
		return "", fmt.Errorf("unable to find ORs: %w", err)
	}
	var ors []string
	for _, loc := range locations {
		start, end := loc[0], loc[1]
		content := s[start:end]
		// handle ANDs
		locs, err := findANDs(content, -1)
		if err != nil {
			return "", fmt.Errorf("unable to find ANDs: %w", err)
		}
		var ands []string
		for _, l := range locs {
			start, end = l[0], l[1]
			content := content[start:end]
			// handle parentheses
			parentheses, err := findOuterParentheses(content, -1)
			if err != nil {
				return "", fmt.Errorf("unable to find parentheses: %w", err)
			}
			for _, p := range parentheses {
				start, end := p[0], p[1]
				content := content[start+1 : end]
				// handle nested
				replacement, err := parser.Process(content)
				if err != nil {
					return "", err
				}
				ands = append(ands, replacement)
			}
			if len(parentheses) > 0 {
				// location is already fully handled
				continue
			}
			// if no parentheses, it should be an operation
			operator := reOperator.FindString(content)
			key := reKey.FindString(content)
			value := reValue.FindString(content)
			if operator == "" || key == "" || value == "" {
				return "", fmt.Errorf("incomplete operation '%s'", content)
			}
			// run key transformers
			for _, t := range parser.keyTransformers {
				key = t(key)
			}
			// check if key is allowed
			if containsString(opts.forbiddenKeys, key) {
				return "", fmt.Errorf("given key '%s' is not allowed", key)
			}
			if len(opts.allowedKeys) > 0 && !containsString(opts.allowedKeys, key) {
				return "", fmt.Errorf("given key '%s' is not allowed", key)
			}
			// parse operation
			var res string
			for _, op := range parser.operators {
				if operator == op.Operator {
					res = op.Formatter(key, value)
					break
				}
			}
			if res == "" {
				return "", fmt.Errorf("unknown operator '%s' in '%s'", operator, content)
			}
			ands = append(ands, res)
		}
		// replacement for AND-block
		replacement := parser.andFormatter(ands)
		ors = append(ors, replacement)
	}
	// replace OR-block and return
	return parser.orFormatter(ors), nil
}

// encodeSpecial encodes all the special strings
// that could get in the way of the parser.
func encodeSpecial(s string) string {
	for dec, enc := range specialEncode {
		s = strings.ReplaceAll(s, dec, enc)
	}
	return s
}

// decodeSpecial decodes all the special strings
// that could get in the way of the parser.
func decodeSpecial(s string) string {
	for dec, enc := range specialEncode {
		s = strings.ReplaceAll(s, enc, dec)
	}
	return s
}

// findParts finds the locations of separated blocks while considering parentheses.
// If n is greater than 0, n parts (from the left) are returned at most.
func findParts(s string, n int, separators ...string) ([][]int, error) {
	if len(s) == 0 {
		return nil, nil
	}
	// validations
	if len(separators) == 0 {
		return nil, fmt.Errorf("no separators given")
	}
	for _, sep := range separators {
		if s[0:1] == sep {
			return nil, fmt.Errorf("given string starts with a separators")
		}
		if s[len(s)-1:] == sep {
			return nil, fmt.Errorf("given string ends with a separators")
		}
	}
	var res [][]int
	var start, par, found int
	runes := []rune(s)
	for i, r := range runes {
		c := string(r)
		// parentheses
		if c == "(" {
			par++
		}
		if c == ")" {
			par--
			if par < 0 {
				return nil, fmt.Errorf("parentheses mismatch")
			}
		}
		// while par for parentheses is not zero,
		// don't bother checking if separators was found
		if par == 0 {
			for _, sep := range separators {
				if c == sep {
					// found part
					found++
					res = append(res, []int{start, i})
					start = i + 1
					if n > 0 && found == n {
						// return if found enough parts
						return res, nil
					}
				}
			}

		}
	}
	// append part after last separators
	end := len(s)
	if start < end {
		res = append(res, []int{start, end})
	}
	return res, nil
}

// findORs finds the locations of all OR blocks in the given string.
// Every location will have two integers, representing the start and end of the block.
// If n is greater than 0, n locations (from the left) are returned at most.
func findORs(s string, n int) ([][]int, error) {
	return findParts(s, n, ",")
}

// findANDs finds the locations of all ANDs blocks in the given string.
// Every location will have two integers, representing the start and end of the block.
// If n is greater than 0, n locations (from the left) are returned at most.
func findANDs(s string, n int) ([][]int, error) {
	return findParts(s, n, ";")
}

// findOuterParentheses finds indexes of opening and closing parentheses.
// Every entry will have two integers, the first one providing the index of the
// opening parentheses, the second one the index of the closing parentheses.
func findOuterParentheses(s string, n int) ([][]int, error) {
	if strings.Count(s, "(") != strings.Count(s, ")") {
		return nil, fmt.Errorf("number of opening and closing parentheses dont match in string '%s'", s)
	}
	var res [][]int
	var start, par, nested, found int
	var op bool
	runes := []rune(s)
	for i, r := range runes {
		c := string(r)
		// start or part of operator
		if c == "=" || c == "!" {
			op = true
		}
		// end of operation
		if (c == "," || c == ";") && nested == 0 {
			op = false
		}
		// opening
		if c == "(" {
			if op {
				nested++
			} else {
				if par == 0 {
					start = i
				}
				par++
			}
		}
		// closing
		if c == ")" {
			if nested > 0 {
				nested--
				if nested < 0 {
					return nil, fmt.Errorf("parentheses mismatch")
				}
				continue
			} else {
				par--
			}
			if par > 0 {
				// we need to find more
				continue
			}
			if par < 0 {
				return nil, fmt.Errorf("parentheses mismatch")
			}
			// found outer parentheses
			found++
			op = false
			res = append(res, []int{start, i})
			start = i + 1
			if n > 0 && found == n {
				// return if found enough parts
				return res, nil
			}

		}
	}
	return res, nil
}
